nighthawk/test/benchmark_http_client_test.cc

```diff
@@ -0,0 +1,289 @@
+                              uint64_t connection_limit, bool use_https, bool use_h2,
+                              uint64_t amount_of_request) {
+    setupBenchmarkClient(uriPath, use_https, use_h2);
+
+    client_->set_connection_timeout(10s);
+    client_->set_max_pending_requests(max_pending);
+    client_->set_connection_limit(connection_limit);
+    EXPECT_TRUE(client_->initialize(runtime_));
+
+    uint64_t amount = amount_of_request;
+    uint64_t inflight_response_count = 0;
+
+    std::function<void()> f = [this, &inflight_response_count]() {
+      --inflight_response_count;
+      if (inflight_response_count == 0) {
+        dispatcher_->exit();
+      }
+    };
+
+    for (uint64_t i = 0; i < amount; i++) {
+      if (client_->tryStartOne(f)) {
+        inflight_response_count++;
+      }
+    }
+
+    EXPECT_EQ(max_pending, inflight_response_count);
+
+    dispatcher_->run(Envoy::Event::Dispatcher::RunType::Block);
+
+    EXPECT_EQ(0, client_->stream_reset_count());
+  }
+
+  std::string getNonZeroValuedCounters() {
+    Client::CounterFilter filter = [](std::string, uint64_t value) { return value > 0; };
+    return client_->countersToString(filter);
+  }
+
+  Envoy::Thread::ThreadFactoryImplPosix thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  Envoy::Api::Impl api_;
+  Envoy::Event::DispatcherPtr dispatcher_;
+  Envoy::Runtime::RandomGeneratorImpl generator_;
+  ::testing::NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  ::testing::NiceMock<Envoy::Runtime::MockLoader> runtime_;
+  std::unique_ptr<Client::BenchmarkClientHttpImpl> client_;
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersions, BenchmarkClientTest,
+                        // testing::ValuesIn(Envoy::TestEnvironment::getIpVersionsForTest()),
+                        testing::ValuesIn({Envoy::Network::Address::IpVersion::v4}),
+                        Envoy::TestUtility::ipTestParamsToString);
+
+TEST_P(BenchmarkClientTest, BasicTestH1) {
+  testBasicFunctionality("/lorem-ipsum-status-200", 1, 1, false, false, 10);
+  EXPECT_EQ("client.upstream_cx_http1_total:1\n\
+client.upstream_cx_rx_bytes_total:3625\n\
+client.upstream_cx_total:1\n\
+client.upstream_cx_tx_bytes_total:82\n\
+client.upstream_rq_pending_total:1\n\
+client.upstream_rq_total:1",
+            getNonZeroValuedCounters());
+}
+
+TEST_P(BenchmarkClientTest, BasicTestH1404) {
+  testBasicFunctionality("/lorem-ipsum-status-404", 1, 1, false, false, 10);
+  EXPECT_EQ("client.upstream_cx_http1_total:1\n\
+client.upstream_cx_protocol_error:1\n\
+client.upstream_cx_rx_bytes_total:97\n\
+client.upstream_cx_total:1\n\
+client.upstream_cx_tx_bytes_total:82\n\
+client.upstream_rq_pending_total:1\n\
+client.upstream_rq_total:1",
+            getNonZeroValuedCounters());
+}
+
+TEST_P(BenchmarkClientTest, BasicTestHttpsH1) {
+  testBasicFunctionality("/lorem-ipsum-status-200", 1, 1, true, false, 10);
+  EXPECT_EQ("client.ssl.ciphers.ECDHE-RSA-AES128-GCM-SHA256:1\n\
+client.ssl.curves.X25519:1\n\
```

Comment:
Yeah, I reckon if you get back a proto here, it would be super easy to write individual EXPECTs or do a proto compare.

Reply:
I switched this to a std::map, I'm planning to consolidate the translation to protoc into a single layer. SG?